src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc

```diff
@@ -0,0 +1,774 @@
+        if (result->payload["topic_filter"].empty()) {
+          result->payload["topic_filter"] += std::string(topic_filter);
+        } else {
+          result->payload["topic_filter"] += ", " + std::string(topic_filter);
+        }
+        PX_ASSIGN_OR_RETURN_INVALID(subscription_options, decoder->ExtractBEInt<uint8_t>());
+        result->payload["subscription_options"] +=
+            "{maximum_qos : " + std::to_string(subscription_options & 0x3) +
+            ", no_local : " + std::to_string((subscription_options >> 2) & 0x1) +
+            ", retain_as_published : " + std::to_string((subscription_options >> 3) & 0x1) +
+            ", retain_handling : " + std::to_string((subscription_options >> 4) & 0x3) + "}";
+        payload_length -= (3 + topic_filter_length);
+      }
+      return ParseState::kSuccess;
+    }
+    case MqttControlPacketType::UNSUBSCRIBE: {
+      size_t payload_length;
+      uint16_t topic_filter_length;
+
+      if ((result->header_fields.find("remaining_length") == result->header_fields.end()) ||
+          (result->header_fields.find("variable_header_length") == result->header_fields.end())) {
+        return ParseState::kInvalid;
+      }
+
+      result->payload["topic_filter"] = "";
+      payload_length = result->header_fields["remaining_length"] -
+                       result->header_fields["variable_header_length"];
+      while (payload_length > 0) {
+        PX_ASSIGN_OR_RETURN_INVALID(topic_filter_length, decoder->ExtractBEInt<uint16_t>());
+        PX_ASSIGN_OR_RETURN_INVALID(std::string_view topic_filter,
+                                    decoder->ExtractString(topic_filter_length));
+        if (result->payload["topic_filter"].empty()) {
+          result->payload["topic_filter"] += std::string(topic_filter);
+        } else {
+          result->payload["topic_filter"] += ", " + std::string(topic_filter);
+        }
+        payload_length -= (2 + topic_filter_length);
+      }
+      return ParseState::kSuccess;
+    }
+    case MqttControlPacketType::SUBACK:
+    case MqttControlPacketType::UNSUBACK: {
+      size_t payload_length;
+      uint8_t reason_code;
+
+      if ((result->header_fields.find("remaining_length") == result->header_fields.end()) ||
+          (result->header_fields.find("variable_header_length") == result->header_fields.end())) {
+        return ParseState::kInvalid;
+      }
+
+      result->payload["reason_code"] = "";
+      payload_length = result->header_fields["remaining_length"] -
+                       result->header_fields["variable_header_length"];
+      while (payload_length > 0) {
+        PX_ASSIGN_OR_RETURN_INVALID(reason_code, decoder->ExtractBEInt<uint8_t>());
+        if (result->payload["reason_code"].empty()) {
+          result->payload["reason_code"] += std::to_string(reason_code);
+        } else {
+          result->payload["reason_code"] += ", " + std::to_string(reason_code);
+        }
+        payload_length -= 1;
+      }
```

Comment:
Wouldn't the else statement be executed once result->payload["reason_code"] is filled with one reason code? The UNSUBACK control type packet can contain multiple reason codes in its payload, and I took this approach to form a readable string as value for payload["reason_code"]. It would start with the first reason code when the string is empty and then concatenate the rest of them to this string.